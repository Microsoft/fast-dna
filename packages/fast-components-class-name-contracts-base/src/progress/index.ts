/**
 * The class name contract for the progress bar component
 */
export interface IProgressClassNameContract {
    progress: string;
}
