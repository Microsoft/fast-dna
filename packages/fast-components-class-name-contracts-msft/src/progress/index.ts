import { ProgressClassNameContract as BaseProgressClassNameContract } from "@microsoft/fast-components-class-name-contracts-base";

/**
 * The class name contract for the progress component
 */
export interface ProgressClassNameContract extends BaseProgressClassNameContract {
    /**
     * The root of the circular progress component
     */
    progressCircular?: string;

    /**
     * The progress size small modifier
     */
    progressCircular__small: string;

    /**
     * The progress size medium modifier
     */
    progressCircular__medium: string;

    /**
     * The progress size large modifier
     */
    progressCircular__large: string;

    /**
     * The progress value indicator
     */
    progress_valueIndicator?: string;

    /**
     * The circular progress value indicator
     */
    progressCircular_valueIndicator?: string;

    /**
     * The indeterminate circular progress value indicator
     */
    progressCircular_valueIndicator__indeterminate?: string;

    /**
     * The indeterminate progress indicator
     */
    progress_indicator?: string;
    
    /**
     * The indeterminate circular progress indicator
     */
    progressCircular_indicator?: string;

    /**
     * The determinate progress value indicator
     */
    progress_indicator__determinate?: string;

    /**
     * The indeterminate progress indicator
     */
    progress_dot?: string;

    /**
     * The indeterminate progress indicator 1 modifier
     */
    progress_dot__1?: string;

    /**
     * The indeterminate progress indicator 2 modifier
     */
    progress_dot__2?: string;

    /**
     * @Deprecated 3.4.0
     */
    progress_dot__3?: string;

    /**
     * @Deprecated 3.4.0
     */
    progress_dot__4?: string;

    /**
     * @Deprecated 3.4.0
     */
    progress_dot__5?: string;
}
