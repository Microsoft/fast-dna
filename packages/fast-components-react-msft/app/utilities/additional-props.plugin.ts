import React from "react";
import { ChildOptionItem, Plugin, PluginProps } from "@microsoft/fast-tooling-react";

export default class MapChildrenPropToCallbackPassingClassName extends Plugin<
    PluginProps
> {
    public resolver(data: any, childOption?: ChildOptionItem): any {
        return (
            className: string,
            disabled?: boolean,
            appearance?: string
        ): React.ReactNode => {
            const callbackArguments: any = {
                className,
            };

            if (callbackArguments.disabled) {
                callbackArguments.disabled = disabled;
            }

            if (callbackArguments.appearance) {
                callbackArguments.appearance = appearance;
            }

            return React.createElement(
                childOption.component,
                Object.assign({}, data, callbackArguments)
            );
        };
    }
}
