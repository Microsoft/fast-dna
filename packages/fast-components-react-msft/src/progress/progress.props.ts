import React from "react";
import {
    ProgressHandledProps as BaseProgressHandledProps,
    ProgressManagedClasses as BaseProgressManagedClasses,
    ProgressUnhandledProps as BaseProgressUnhandledProps,
} from "@microsoft/fast-components-react-base";
import {
    ManagedClasses,
    ProgressClassNameContract,
} from "@microsoft/fast-components-class-name-contracts-msft";
import { Subtract } from "utility-types";

export enum ProgressSize {
    _16 = 16,
    _32 = 32,
    _64 = 64,
}

/* tslint:disable:no-empty-interface */
export interface ProgressManagedClasses
    extends ManagedClasses<ProgressClassNameContract> {}
export interface ProgressHandledProps
    extends Subtract<BaseProgressHandledProps, BaseProgressManagedClasses>,
        ProgressManagedClasses {
    circular?: boolean;
    size?: ProgressSize;
}

export interface ProgressUnhandledProps extends BaseProgressHandledProps {}
/* tslint:enable:no-empty-interface */
export type ProgressProps = ProgressHandledProps & ProgressUnhandledProps;
