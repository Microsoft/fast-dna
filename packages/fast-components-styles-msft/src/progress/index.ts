import designSystemDefaults, { IDesignSystem } from "../design-system";
import { ComponentStyles, ComponentStyleSheet, ICSSRules } from "@microsoft/fast-jss-manager";
import { applyLocalizedProperty, Direction, toPx } from "@microsoft/fast-jss-utilities";
import { IMSFTProgressClassNameContract } from "@microsoft/fast-components-class-name-contracts-msft";
import Chroma from "chroma-js";
import { get } from "lodash";

/* tslint:disable-next-line */
const styles: ComponentStyles<IMSFTProgressClassNameContract, IDesignSystem> = (config: IDesignSystem): ComponentStyleSheet<IMSFTProgressClassNameContract, IDesignSystem> => {
    const backgroundColor: string = get(config, "backgroundColor") || designSystemDefaults.backgroundColor;
    const brandColor: string = get(config, "brandColor") || designSystemDefaults.brandColor;
    const direction: Direction = get(config, "direction") || designSystemDefaults.direction;
    const foregroundColor: string = get(config, "foregroundColor") || designSystemDefaults.foregroundColor;

    return {
        progress: {
            display: "inline-block",
            width: "100%"
        },
        progress_indicator__determinate: {
            display: "flex",
            background: Chroma(foregroundColor).alpha(0.2).css(),
            height: "4px",
            borderRadius: "2px",
            width: "100%"
        },
        progress_indicator__determinate_bar: {
            display: "inline-block",
            background: brandColor,
            borderRadius: "2px",
            height: "100%"
        },
        progress_indicator__indeterminate: {
            position: "relative",
            display: "flex",
            alignItems: "center",
            height: "10px",
            width: "100%"
        },
        progress_dot__base: {
            position: "absolute",
            display: "inline-block",
            opacity: "0",
            width: "100%",
            height: "5px",
            borderRadius: "50%",
            animation: "dots 4s infinite",
            "&::after": {
                content: "''",
                position: "absolute",
                display: "inline-block",
                background: brandColor,
                width: "5px",
                height: "5px",
                borderRadius: "50%"
            }
        },
        progress_dot__1: {
            animationDelay: "0.25s",
            left: "-20px"
        },
        progress_dot__2: {
            animationDelay: "0.2s",
            left: "-10px"
        },
        progress_dot__3: {
            animationDelay: "0.15s",
            left: "0px"
        },
        progress_dot__4: {
            animationDelay: "0.1s",
            left: "10px"
        },
        progress_dot__5: {
            animationDelay:" 0.05s",
            left: "20px"
        },
        "@keyframes dots": {
            "0%": {},
            "20%": {
                transform: "translateX(0)",
                animationTimingFunction: "ease-out",
                opacity: "0",
            },
            "25%": {
                opacity: "1",
            },
            "35%": {
                transform: "translateX(45%)",
                animationTimingFunction: "linear",
            },
            "65%": {
                transform: "translateX(55%)",
                animationTimingFunction: "ease-out"
            },
            "75%": {
                opacity: "1",
            },
            "80%": {
                transform: "translateX(100%)",
                opacity: "0",
            },
            "100%": {
            }
        }
    };
};

export default styles
