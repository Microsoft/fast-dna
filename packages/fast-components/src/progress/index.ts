export * from "./base-progress";
export * from "./circular-progress";
export * from "./linear-progress";
