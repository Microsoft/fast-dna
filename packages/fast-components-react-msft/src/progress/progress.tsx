import * as React from "react";
import * as ReactDOM from "react-dom";
import { get } from "lodash-es";
import { Foundation, HandledProps } from "@microsoft/fast-components-react-base";
import { IProgressManagedClasses, IProgressUnhandledProps, ProgressProps } from "./progress.props";
import { IManagedClasses, IMSFTProgressClassNameContract } from "@microsoft/fast-components-class-name-contracts-msft";
import { Progress as BaseProgress, ProgressType } from "@microsoft/fast-components-react-base";

/* tslint:disable-next-line */
class Progress extends Foundation<ProgressProps & IManagedClasses<IMSFTProgressClassNameContract>,  React.AllHTMLAttributes<HTMLElement>, {}> {
    public static defaultProps: Partial<ProgressProps> = {
        minValue: 0,
        maxValue: 100
    };

    protected handledProps: HandledProps<ProgressProps & IManagedClasses<IMSFTProgressClassNameContract>> = {
        children: void 0,
        value: void 0,
        minValue: void 0,
        maxValue: void 0,
        managedClasses: void 0,
    };

    /**
     * Renders the component
     */
    public render(): React.ReactElement<HTMLButtonElement | HTMLAnchorElement> {
        return (
            <BaseProgress
                {...this.unhandledProps()}
                className={this.generateClassNames()}
                managedClasses={this.props.managedClasses}
                value={this.props.value}
                minValue={this.props.minValue}
                maxValue={this.props.maxValue}
            >
                <div
                    className={this.props.managedClasses.progress_indicator__determinate}
                    slot={ProgressType.determinate}
                >
                    <div
                        className={this.props.managedClasses.progress_indicator__determinate_bar}
                        style={{width: `${this.props.value}%`}}
                    />
                </div>
                <div
                    slot={ProgressType.indeterminate}
                    className={this.props.managedClasses.progress_indicator__indeterminate}
                >
                    {this.renderProgressSpans(this.generateProgressSpans())}
                </div>
            </BaseProgress>
        );
    }

    public generateProgressSpans(): Array<{id: number}> {
        return [{id: 1}, {id: 2}, {id: 3}, {id: 4}, {id: 5}];
    }

    /**
     * Generate indeterminate spans
     */
    private renderProgressSpans(spanData: any[]): JSX.Element[] {
        return spanData.map((span: any, index: number) => {
            let className: string = this.props.managedClasses.progress_dot__base;
            className = `${className} ${this.props.managedClasses[`progress_dot__${span.id}`]}`;
            return (
                <span
                    className={className}
                    key={span.id}
                />
            );
        });
    }
}

export default Progress;
export * from "./progress.props";
export { IMSFTProgressClassNameContract };
