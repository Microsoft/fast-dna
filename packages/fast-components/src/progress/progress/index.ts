import { customElement } from "@microsoft/fast-element";
import { BaseProgress } from "../base-progress";
import { ProgressTemplate as template } from "./progress.template";
import { ProgressStyles as styles } from "./progress.styles";
import { consumer } from "../../design-system-provider/design-system-consumer";
import {
    neutralfillrest,
    accentfillrest,
    neutralforegroundhint,
} from "../../styles/recipes";

@customElement({
    name: "fast-progress",
    template,
    styles,
})
@consumer({
    recipes: [neutralfillrest, accentfillrest, neutralforegroundhint],
})
export class FASTProgress extends BaseProgress {}
export * from "./progress.template";
export * from "./progress.styles";
