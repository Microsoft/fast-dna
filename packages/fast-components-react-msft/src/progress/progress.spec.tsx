import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { configure, mount, shallow } from "enzyme";
import MSFTProgress, { ProgressClassNameContract } from "./progress";
import { ProgressProps } from "./index";
import { DisplayNamePrefix } from "../utilities";
import { ProgressSize } from "./progress.props";

/*
 * Configure Enzyme
 */
configure({ adapter: new Adapter() });

const managedClasses: ProgressClassNameContract = {
    progress: "progress-class",
    progressCircular: "progressCircular-class",
    progressCircular__small: "progressCircular__small-class",
    progressCircular__medium: "progressCircular__medium-class",
    progressCircular__large: "progressCircular__large-class",
    progress_valueIndicator: "progress_valueIndicator-class",
    progressCircular_valueIndicator: "progressCircular_valueIndicator-class",
    progressCircular_valueIndicator__indeterminate:
        "progressCircular_valueIndicator__indeterminate-class",
    progress_indicator: "progress_indicator-class",
    progressCircular_indicator: "progressCircular_indicator-class",
    progress_indicator__determinate: "progress_indicator__determinate-class",
    progress_dot: "progress_dot-class",
    progress_dot__1: "progress_dot__1-class",
    progress_dot__2: "progress_dot__2-class",
};

describe("progress", (): void => {
    test("should have a displayName that matches the component name", () => {
        expect(`${DisplayNamePrefix}${(MSFTProgress as any).name}`).toBe(
            MSFTProgress.displayName
        );
    });

    test("should not throw if managedClasses are not provided", () => {
        expect(() => {
            shallow(<MSFTProgress />);
        }).not.toThrow();
    });

    test("should use circular class names when circular prop is set to true", () => {
        const rendered: any = mount(
            <MSFTProgress managedClasses={managedClasses} circular={true} />
        );

        expect(rendered.exists("div.progressCircular-class")).toBe(true);
        expect(rendered.exists("circle.progressCircular_valueIndicator-class")).toBe(
            true
        );
        expect(
            rendered.exists("circle.progressCircular_valueIndicator__indeterminate-class")
        ).toBe(true);
    });

    test("should use not use interdeterminate class names when value prop is set", () => {
        const rendered: any = shallow(
            <MSFTProgress managedClasses={managedClasses} circular={true} value={75} />
        );

        expect(
            rendered.exists("circle.progressCircular_valueIndicator__indeterminate-class")
        ).toBe(false);
    });

    test("should render the correct `size` when `size` prop is small", () => {
        const rendered: any = mount(
            <MSFTProgress
                managedClasses={managedClasses}
                circular={true}
                size={ProgressSize.small}
            />
        );

        expect(rendered.find("svg").prop("className")).toContain("small");
    });

    test("should render the correct `size` when `size` prop is large", () => {
        const rendered: any = mount(
            <MSFTProgress
                managedClasses={managedClasses}
                circular={true}
                size={ProgressSize.large}
            />
        );

        expect(rendered.find("svg").prop("className")).toContain("large");
    });

    test("should render default size if none is specified", () => {
        const rendered: any = mount(
            <MSFTProgress managedClasses={managedClasses} circular={true} />
        );

        expect(rendered.find("svg").prop("className")).toContain("medium");
    });
});
