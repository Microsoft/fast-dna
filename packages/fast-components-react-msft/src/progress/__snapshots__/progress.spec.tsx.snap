// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`progress snapshots Progress: 0 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={50}
  className="progress-0-1-11"
  role="progressbar"
>
  <div
    className="progress_indicator__determinate-0-1-12"
    slot="determinate"
  >
    <div
      className="progress_indicator__determinate_bar-0-1-13"
      style={
        Object {
          "width": "50%",
        }
      }
    />
  </div>
</div>
`;

exports[`progress snapshots Progress: 1 1`] = `
<div
  aria-valuemax={100}
  aria-valuemin={0}
  aria-valuenow={undefined}
  className="progress-0-1-31"
  role="progressbar"
>
  <div
    className="progress_indicator__indeterminate-0-1-34"
    slot="indeterminate"
  >
    <span
      className="progress_dot__base-0-1-35 progress_dot__1-0-1-36"
    />
    <span
      className="progress_dot__base-0-1-35 progress_dot__2-0-1-37"
    />
    <span
      className="progress_dot__base-0-1-35 progress_dot__3-0-1-38"
    />
    <span
      className="progress_dot__base-0-1-35 progress_dot__4-0-1-39"
    />
    <span
      className="progress_dot__base-0-1-35 progress_dot__5-0-1-40"
    />
  </div>
</div>
`;
